desc/singularities.py

```diff
@@ -0,0 +1,758 @@
+            src_zeta, axis=1
+        ), "singular integration requires rectangular source grid in theta and zeta"
+        assert islinspaced(
+            src_theta, axis=1
+        ), "singular integration requires source nodes be equally spaced in theta"
+        assert islinspaced(
+            src_zeta, axis=0
+        ), "singular integration requires source nodes be equally spaced in zeta"
+        assert np.all(src_grid.nodes[:, 0] == eval_grid.nodes[0, 0]) and np.all(
+            eval_grid.nodes[:, 0] == src_grid.nodes[0, 0]
+        ), "singular integration requires source and eval grids on the same surface."
+
+        self._eval_grid = eval_grid
+        self._src_grid = src_grid
+        self._s = s
+        self._q = q
+
+        r, w, dr, dw = _get_quadrature_nodes(q)
+
+        src_dtheta = src_grid.spacing[:, 1]
+        src_dzeta = src_grid.spacing[:, 2] / src_grid.NFP
+        eval_theta = eval_grid.nodes[:, 1]
+        eval_zeta = eval_grid.nodes[:, 2]
+
+        h_t = jnp.mean(src_dtheta)
+        h_z = jnp.mean(src_dzeta)
+
+        theta_q = eval_theta[:, None] + s / 2 * h_t * r * jnp.sin(w)
+        zeta_q = eval_zeta[:, None] + s / 2 * h_z * r * jnp.cos(w)
+
+        basis = DoubleFourierSeries(M=src_grid.M, N=src_grid.N, NFP=src_grid.NFP)
+        A = basis.evaluate(src_grid.nodes)
+        Ainv = jnp.linalg.pinv(A, rcond=None)
+
+        B = jnp.zeros((*theta_q.shape, basis.num_modes))
+
+        def body(i, B):
+            x = jnp.array(
+                [
+                    jnp.zeros_like(theta_q[i]),
+                    theta_q[i],
+                    zeta_q[i],
+                ]
+            ).T
+            Bi = basis.evaluate(jnp.atleast_2d(x))
+            B = put(B, i, Bi)
+            return B
+
+        B = fori_loop(0, B.shape[0], body, B)
+        self._mat = B @ Ainv
+
+    def __call__(self, f, i):
+        """Interpolate data to polar grid points.
+
+        Parameters
+        ----------
+        f : ndarray
+            Data at source grid points to interpolate.
+        i : int
+            Index of polar node.
+
+        Returns
+        -------
+        fi : ndarray
+            Source data interpolated to ith polar node.
+        """
+        return self._mat[:, i] @ f
+
+
+def _chi(rho):
+    """Partition of unity function. Eq 39 in [2]."""
+    return jnp.exp(-36 * jnp.abs(rho) ** 8)
+
+
+def _rho(theta, zeta, theta0, zeta0, dtheta, dzeta, s):
+    """Polar grid radial coordinate. Argument of Chi in eq. 36 in [2]."""
+    dt = abs(theta - theta0)
+    dz = abs(zeta - zeta0)
+    dt = jnp.minimum(dt, 2 * np.pi - dt)
+    dz = jnp.minimum(dz, 2 * np.pi - dz)
+    return 2 / s * jnp.sqrt((dt / dtheta) ** 2 + (dz / dzeta) ** 2)
+
+
+def _nonsingular_part(eval_data, eval_grid, src_data, src_grid, s, kernel, loop=False):
+    """Integrate kernel over non-singular points.
+
+    Generally follows sec 3.2.1 of [2].
+    """
+    assert src_grid.NFP == int(src_grid.NFP)
```

Comment:
I'm getting errors here when NFP is of type `numpy.int64`

Reply:
Can you provide an example? That should work fine for any integer type

Reply:
Actually it is initially type `numpy.int64` but then during the step: 
`Compiling objective function and derivatives: ['Boundary error']`
it somehow becomes type `numpy.ndarray` and that throws the error. 

I'm trying to reproduce it with an equilibrium I can share...

Reply:
did you figure out the source of this issue?